File: src/main/scala/temportalist/compression/main/common/Effects.scala
package temportalist.compression.main.common

import java.util

import net.minecraft.entity.Entity
import net.minecraft.entity.item.EntityItem
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.ItemStack
import net.minecraft.util.math.AxisAlignedBB
import net.minecraft.util.{ResourceLocation, SoundCategory, SoundEvent}
import net.minecraft.world.World
import net.minecraftforge.event.entity.player.EntityItemPickupEvent
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
import temportalist.compression.main.common.init.Compressed
import temportalist.compression.main.common.item.ItemDenseArmor
import temportalist.compression.main.common.lib.EnumTier
import temportalist.origin.api.common.lib.Vect

import scala.collection.JavaConversions
import scala.collection.mutable.ListBuffer
import scala.util.control.Breaks._

/**
  *
  * Created by TheTemportalist on 4/15/2016.
  *
  * @author TheTemportalist
  */
object Effects {

	def shouldUseTickingTile(itemStack: ItemStack): Boolean = {
		this.canUseBlackHole(itemStack)
	}

	val portalSound = new ResourceLocation("entity.endermen.teleport")

	def getTierNumber(itemStack: ItemStack): Int = Compressed.getTier(itemStack).ordinal()

	def canUse(itemStack: ItemStack, tierOrdinal: Int): Boolean =
		this.getTierNumber(itemStack) >= tierOrdinal - 1

	def canUseCompressor(itemStack: ItemStack): Boolean = this.canUse(itemStack, Options.compressor)

	def canUseMagnetI(itemStack: ItemStack): Boolean = this.canUse(itemStack, Options.magnetI)

	def canUseAttractionI(itemStack: ItemStack): Boolean = this.canUse(itemStack, Options.attractionI)

	def canUseMagnetII(itemStack: ItemStack): Boolean = this.canUse(itemStack, Options.magnetII)

	def canUseAttractionII(itemStack: ItemStack): Boolean = this.canUse(itemStack, Options.attractionII)

	def canUseAttractionIII(itemStack: ItemStack): Boolean = this.canUse(itemStack, Options.attractionIII)

	def canUseBlackHole(itemStack: ItemStack): Boolean = this.canUse(itemStack, Options.blackHole)

	def canUseAny(itemStack: ItemStack, set: Int*): Boolean = {
		set.foreach(tierOrdinalPlus => if (this.canUse(itemStack, tierOrdinalPlus)) return true)
		false
	}

	def getLowestTier(tiers: Int*): Int = {
		var lowest: Int = -1
		tiers.foreach(tier => if (tier >= 0 && (lowest < 0 || tier < lowest)) lowest = tier)
		lowest
	}

	def getLowestTierMagnet: Int = this.getLowestTier(Options.magnetI, Options.magnetI)

	def iterateOverPlayerInventoryForSample(player: EntityPlayer, sample: ItemStack,
			onSampleFound: (Int, ItemStack, Boolean) => Boolean, min: Int = 9, max: Int = 36): Unit = {
		breakable {
			for (index <- min until max) {
				// just the storage part
				val stackInSlot = player.inventory.getStackInSlot(index)
				if (stackInSlot != null) {
					val slotSample = Compressed.getSampleFromUnknown(stackInSlot)
					if (slotSample.getItem == sample.getItem &&
							slotSample.getItemDamage == sample.getItemDamage) {
						if (onSampleFound(index, stackInSlot, Compressed.isCompressed(stackInSlot)))
							break()
					}
				}
			}
		}
	}

	def appendToInventory(player: EntityPlayer, stack: ItemStack, min: Int = 9, max: Int = 36): Boolean = {
		for (index <- min until max) {
			val stackInSlot = player.inventory.getStackInSlot(index)
			if (stackInSlot == null) {
				player.inventory.setInventorySlotContents(index, stack)
				return true
			}
		}
		false
	}

	@SubscribeEvent
	def onItemPickup(event: EntityItemPickupEvent): Unit = {
		val player = event.getEntityPlayer
		val entityItem = event.getItem
		if (player == null || entityItem == null) return

		val entityStack = entityItem.getEntityItem
		var entityStackSample: ItemStack = null
		var entityStackSize: Long = entityStack.stackSize
		if (Compressed.isCompressed(entityStack)) {
			entityStackSample = Compressed.getSampleStack(entityStack)
			entityStackSize *= Compressed.getSize(entityStack)
		}
		else {
			entityStackSample = entityStack.copy()
			entityStackSample.stackSize = 1
		}

		var indexCompressor: Int = -1
		var stackCompressor: ItemStack = null
		this.iterateOverPlayerInventoryForSample(player, entityStackSample,
			(index: Int, stack: ItemStack, isCompressed: Boolean) => {
				if (isCompressed && this.canUseCompressor(stack)) {
					indexCompressor = index
					stackCompressor = stack
					true
				}
				else false
			}: Boolean, min = 0, max = 36 + 4 + 1 // 36 = main, 4 = armor, 1 = offhand
		)

		if (indexCompressor >= 0) {
			event.setCanceled(true)
			event.getItem.setDead()
			this.addToAndCompressInventory(player, entityStackSize, entityStackSample)
			player.inventory.markDirty()
		}

	}

	private def addToAndCompressInventory(player: EntityPlayer,
			excessSize: Long, sample: ItemStack): Unit = {
		var size: Long = excessSize

		this.iterateOverPlayerInventoryForSample(player, sample,
			(index: Int, stack: ItemStack, isCompressed: Boolean) => {
				size += stack.stackSize * (if (isCompressed) Compressed.getSize(stack) else 1)
				player.inventory.removeStackFromSlot(index)
				false
			}: Boolean
		)

		val stackList = this.getStackListFromSize(size, sample)

		for (stack <- stackList) {
			if (!this.appendToInventory(player, stack))
				player.dropItem(stack, false, false)
		}

	}

	def getStackListFromSize(sizeIn: Long, sample: ItemStack): ListBuffer[ItemStack] = {
		var size = sizeIn

		val sizeMap = new util.EnumMap[EnumTier, Int](classOf[EnumTier])
		def append(tier: EnumTier, quantity: Int): Unit =
			sizeMap.put(tier, quantity + (if (sizeMap.containsKey(tier)) sizeMap.get(tier) else 0))

		while (size > EnumTier.getTail.getSizeMax) {
			append(EnumTier.getTail, 1)
			size -= EnumTier.getTail.getSizeMax
		}

		while (size > 8) {
			var tier = EnumTier.getTierForSize(size)
			if (size - tier.getSizeMax < 0)
				tier = EnumTier.getTier(tier.ordinal() - 1)
			append(tier, 1)
			size -= tier.getSizeMax
		}

		val stackList = ListBuffer[ItemStack]()
		for (entry <- JavaConversions.asScalaSet(sizeMap.entrySet())) {
			val stack = Compressed.create(sample, tier = entry.getKey)
			stack.stackSize = entry.getValue
			stackList += stack
		}
		if (size > 0) {
			val stack = sample.copy()
			stack.stackSize = size.toInt
			stackList += stack
		}

		stackList
	}

	def onInvUpdateCompressed(world: World, player: EntityPlayer, stack: ItemStack): Unit = {
		this.onInvUpdateCompressed_Magnet(world, player, stack)
	}

	def onInvUpdateCompressed_Magnet(world: World, player: EntityPlayer, stack: ItemStack): Unit = {
		if (!(this.canUseMagnetI(stack) || this.canUseMagnetII(stack))) return
		val lowestTier = this.getLowestTierMagnet
		if (lowestTier < 0) return

		val stackTierNumber = Compressed.getTier(stack).ordinal() + 1
		val stackSample = Compressed.getSampleStack(stack)

		val radiusFactor = stackTierNumber - lowestTier
		val position = new Vect(player)
		val motion = new Vect(player.motionX, player.motionY, player.motionZ)
		val boundingBoxBase = player.getEntityBoundingBox.expand(1, 0.5, 1)

		val radius = radiusFactor + 1.5D

		this.iterateEntitiesAround(boundingBoxBase, radius, world, player, position, motion,
			// Filter: Should the entity passed by pulled
			(entity: Entity) => {
				entity match {
					case entityItem: EntityItem =>
						if (this.canUseMagnetII(stack)) true
						else this.doesMatchSample(stackSample, entityItem)
					case _ => false
				}
			}: Boolean, null,
			// Post effect: When an entity is pulled
			null
		)
	}

	def onEntityUpdateCompressed(world: World, entity: EntityItem, itemStack: ItemStack): Unit = {
		if (!itemStack.hasTagCompound) return
		val tiers = Seq(Options.attractionI, Options.attractionII, Options.attractionIII)
		if (!this.canUseAny(itemStack, tiers:_*)) return
		val lowestTier = this.getLowestTier(tiers:_*)
		if (lowestTier < 0) return

		val stackTierNumber = Compressed.getTier(itemStack).ordinal() + 1
		val stackSample = Compressed.getSampleStack(itemStack)

		val radiusFactor = stackTierNumber - lowestTier
		val position = new Vect(entity)
		val motion = new Vect(entity.motionX, entity.motionY, entity.motionZ)
		val boundingBoxBase = entity.getEntityBoundingBox
		val radius = 1.5 + radiusFactor

		this.iterateEntitiesAround(boundingBoxBase, radius, world, entity, position, motion,
			// Filter: Should the entity passed by pulled
			(entity: Entity) => {
				entity match {
					case player: EntityPlayer =>
						this.canUseAttractionIII(itemStack) &&
								!player.capabilities.isCreativeMode && !player.isSneaking
					case entityItem: EntityItem =>
						this.canUseAttractionIII(itemStack) ||
								this.canUseAttractionII(itemStack) ||
								this.doesMatchSample(stackSample, entityItem)
					case _ => this.canUseAttractionIII(itemStack)
				}
			}: Boolean,
			(entity: Entity) => {
				entity match {
					case player: EntityPlayer => 0.25D * (4 - ItemDenseArmor.getClothedCount(player))
					case _ => 1
				}
			}: Double,
			// Post effect: When an entity is pulled
			{
				case entityItem: EntityItem =>
					/*
					if (this.doesMatchSample(stackSample, entityItem)) {
						val sizeTotal = Compressed.getSize(itemStack) +
								Compressed.getTotalSizeForUnknown(entityItem.getEntityItem)
						val stackList = this.getStackListFromSize(sizeTotal, stackSample)
						entity.setEntityItemStack(stackList.remove(0))
						if (stackList.nonEmpty) entityItem.setEntityItemStack(stackList.remove(0))
						while (stackList.nonEmpty) {
							val stack = stackList.remove(0)
							world.spawnEntityInWorld(
								if (Compressed.isCompressed(stack))
									new EntityItemCompressed(world, position, motion, stack)
								else {
									val ei = new EntityItem(
										world, position.x, position.y, position.z, stack)
									ei.motionX = motion.x
									ei.motionY = motion.y
									ei.motionZ = motion.z
									ei.setDefaultPickupDelay()
									ei
								}
							)
						}
					}
					*/
				case _ =>
			}
		)

	}

	def iterateEntitiesAround(boundingBoxBase: AxisAlignedBB, radius: Double,
			world: World, entityToExclude: Entity, position: Vect, motion: Vect,
			shouldPull: (Entity) => Boolean, getSpeed: (Entity) => Double,
			onPull: (Entity) => Unit): Unit = {
		val boundingBox = boundingBoxBase.expand(radius, radius, radius)
		val entList = world.getEntitiesWithinAABBExcludingEntity(entityToExclude, boundingBox)
		for (entityBB <- JavaConversions.asScalaBuffer(entList)) {
			if (shouldPull == null || shouldPull(entityBB)) {
				this.pullEntityTowards(entityBB, position, motion,
					speed = if (getSpeed == null) 1 else getSpeed(entityBB))
				if (onPull != null) onPull(entityBB)
			}
		}
	}

	def pullEntityTowards(entityToPull: Entity, pos: Vect, motion: Vect, speed: Double = 1): Unit = {
		if (speed == 0) return
		var distX = pos.x - entityToPull.posX
		var distY = pos.y - entityToPull.posY
		var distZ = pos.z - entityToPull.posZ

		val distance = Math.sqrt(distX * distX + distY * distY + distZ * distZ) * 2

		distX = distX / distance + motion.x * 0.5
		distY = distY / distance + motion.y * 0.5
		distZ = distZ / distance + motion.z * 0.5

		entityToPull.motionX = distX * speed
		entityToPull.motionY = distY * speed
		entityToPull.motionZ = distZ * speed
		entityToPull.isAirBorne = true

		if (entityToPull.isCollidedHorizontally) {
			entityToPull.motionY += 1
		}

		entityToPull.fallDistance = 0f

		if (entityToPull.worldObj.rand.nextInt(20) == 0) {
			val sound = SoundEvent.REGISTRY.getObject(this.portalSound)
			val pitch = 0.85f - entityToPull.worldObj.rand.nextFloat() * 3f / 10f
			entityToPull.getEntityWorld.playSound(
				entityToPull.posX, entityToPull.posY, entityToPull.posZ,
				sound, SoundCategory.BLOCKS, 0.6f, pitch, false
			)
		}
	}

	def doesMatchSample(sample: ItemStack, entityItem: EntityItem): Boolean = {
		val entityItemStack = entityItem.getEntityItem
		val entityItemSample = Compressed.getSampleFromUnknown(entityItemStack)
		sample.getItem == entityItemSample.getItem && sample.getItemDamage == entityItemSample.getItemDamage
	}

}

File: src/main/java/com/temportalist/compression/common/effects/Effects.java
package com.temportalist.compression.common.effects;

import com.temportalist.compression.common.entity.EntityCompressed;
import com.temportalist.compression.common.init.CompressedStack;
import com.temportalist.compression.common.lib.EnumTier;
import net.minecraft.entity.Entity;
import net.minecraft.entity.item.EntityItem;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.entity.player.InventoryPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.util.*;
import net.minecraft.util.math.AxisAlignedBB;
import net.minecraft.util.math.Vec3d;
import net.minecraft.world.World;
import net.minecraftforge.event.entity.player.EntityItemPickupEvent;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;

import javax.annotation.Nullable;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.function.BiFunction;
import java.util.function.Consumer;
import java.util.function.Function;

@Mod.EventBusSubscriber
public class Effects {

    public static ResourceLocation portalSound = new ResourceLocation("entity.endermen.teleport");

    public static boolean canUseAny(ItemStack stack, EnumEffect... effects) {
        if (!CompressedStack.isCompressed(stack)) return false;
        for (EnumEffect effect : effects) {
            if (effect != null && effect.canDoEffect(stack)) return true; // effects tier set via config
        }
        return false;
    }

    public static boolean doTickCompressedBlock(ItemStack stack) {
        return Effects.canUseAny(stack, EnumEffect.BLACK_HOLE);
    }

    @SubscribeEvent
    public static void itemPickup(EntityItemPickupEvent event){
        EntityPlayer player = event.getEntityPlayer();
        EntityItem entity = event.getItem();
        if (player == null || entity == null) return;

        ItemStack entityStack = entity.getItem();
        ItemStack entityStackSample = null;
        EnumTier entityTier = null;
        if (CompressedStack.isCompressed(entityStack)) {
            entityStackSample = CompressedStack.createSampleStack(entityStack);
            entityTier = CompressedStack.getTier(entityStack);
        }
        else {
            entityStackSample = entityStack.copy();
            // tier stays null
        }

        int indexCompressor = Effects.matchItemStack(player, entityStackSample, true,
                (Tuple<ItemStack, Integer> invStack, Boolean isCompressed) ->
                        isCompressed && canUseAny(invStack.getFirst(), EnumEffect.COMPRESSOR)
        );

        if (indexCompressor >= 0) {
            event.setCanceled(true);
            event.getItem().setDead();
            Effects.compressInventory(player, entityStackSample, entityTier);
        }

    }

    public static boolean doesMatch(ItemStack a, ItemStack b) {
        boolean sameItem = a.getItem() == b.getItem();
        boolean sameDamage = a.getItemDamage() == b.getItemDamage();
        boolean sameTag =  ItemStack.areItemStackTagsEqual(a, b);
        return sameItem && sameDamage && sameTag;
    }

    public static int matchItemStack(EntityPlayer player, ItemStack sampleIn, boolean includeHotbar,
                                     @Nullable BiFunction<Tuple<ItemStack, Integer>, Boolean, Boolean> onFound) {
        int slotMin = 0, slotMax =
                        player.inventory.mainInventory.size() +
                        player.inventory.armorInventory.size() +
                        player.inventory.offHandInventory.size();

        for (int slot = slotMin; slot < slotMax; slot++) {
            if (!includeHotbar && InventoryPlayer.isHotbar(slot)) continue;
            ItemStack stackInSlot = player.inventory.getStackInSlot(slot);
            if (stackInSlot != ItemStack.EMPTY) {
                boolean isCompressed = CompressedStack.isCompressed(stackInSlot);
                ItemStack sample = isCompressed ? CompressedStack.createSampleStack(stackInSlot) : stackInSlot.copy();
                if (sample != null && Effects.doesMatch(sample, sampleIn) &&
                        (onFound == null || onFound.apply(new Tuple<>(stackInSlot, slot), isCompressed))
                ) {
                    return slot;
                }
            }
        }
        return -1;
    }

    public static void compressInventory(EntityPlayer player, ItemStack stackIn, EnumTier tierIn) {

        ArrayList<ItemStack> alikeStacks = new ArrayList<>();

        Effects.matchItemStack(player, stackIn, false,
                (Tuple<ItemStack, Integer> invStack, Boolean isCompressed) -> {
                    alikeStacks.add(player.inventory.removeStackFromSlot(invStack.getSecond()));
                    return false;
                }
        );

        long totalCount = (tierIn == null ? 1 : tierIn.getSizeMax()) * stackIn.getCount();
        for (ItemStack stack : alikeStacks) {
            EnumTier tier = CompressedStack.getTier(stack);
            totalCount += stack.getCount() * (tier != null ? tier.getSizeMax() : 1);
        }

        List<ItemStack> stackList = CompressedStack.createStackList(
                CompressedStack.createSampleStack(stackIn), totalCount
        );
        int slot;
        for (ItemStack stack : stackList) {
            slot = Effects.getFirstEmptyStackNonHotbar(player.inventory);
            if (!player.inventory.add(slot, stack)) {
                player.dropItem(stack, true, false);
            }
        }

        player.inventory.markDirty();
    }

    public static int getFirstEmptyStackNonHotbar(InventoryPlayer player) {
        for (int i = InventoryPlayer.getHotbarSize(); i < player.mainInventory.size(); ++i)
        {
            if ((player.mainInventory.get(i)).isEmpty())
            {
                return i;
            }
        }
        return -1;
    }

    public static void attractEntities(AxisAlignedBB boundingBox, float radius, World world, Entity entity,
                                Vec3d position,
                                @Nullable Function<Entity, Boolean> isValid,
                                @Nullable Function<Entity, Float> getSpeed,
                                @Nullable Consumer<Entity> onPull
                                ) {
        boundingBox = boundingBox.grow(radius, radius, radius);
        List<Entity> ents = world.getEntitiesWithinAABBExcludingEntity(entity, boundingBox);
        for (Entity ent : ents ) {
            if (isValid == null || isValid.apply(ent)) {
                float speed  = getSpeed == null ? 1f : getSpeed.apply(ent);
                if (speed <= 0) continue;
                Effects.pullEntityTowards(ent, position, speed);
                if (onPull != null) onPull.accept(ent);
            }
        }
    }

    public static void pullEntityTowards(Entity ent, Vec3d position, float speed) {
        // Vector from player to entity position: position - entity.position
        Vec3d diff = position.subtract(ent.getPositionVector());
        diff = diff.normalize();//.scale(speed);
        diff = diff.scale(speed);

        ent.setVelocity(diff.x, diff.y, diff.z);

        if (ent.isCollidedHorizontally) {
            ent.motionY += 1;
        }

        ent.fallDistance = 0f;

        if (ent.getEntityWorld().rand.nextInt(20) == 0) {
            SoundEvent sound = SoundEvent.REGISTRY.getObject(Effects.portalSound);
            float pitch = 0.85f - ent.getEntityWorld().rand.nextFloat() * 3f / 10f;
            ent.getEntityWorld().playSound(
                    ent.posX, ent.posY, ent.posZ,
                    sound, SoundCategory.BLOCKS, 0.6f, pitch, false
            );
        }
    }

    public static void onUpdateEntityInventory(World world, Entity entity, ItemStack stack) {
        if (entity.isSneaking()) return;
        if (!Effects.canUseAny(stack, EnumEffect.MAGNET_I, EnumEffect.MAGNET_II)) return;
        boolean validMagnetII = Effects.canUseAny(stack, EnumEffect.getLowestTier(EnumEffect.MAGNET_I, EnumEffect.MAGNET_II));

        ItemStack sample = CompressedStack.createSampleStack(stack);
        // radiusFactor is the tier of the stack - the lowest possible tier for a magnet
        float radiusFactor = CompressedStack.getTier(stack).ordinal() - EnumEffect.MAGNET_I.getTier().ordinal();
        float radius = 1.5f + radiusFactor;

        AxisAlignedBB boundingBox = entity.getEntityBoundingBox().expand(1, 0.5, 1);
        Effects.attractEntities(boundingBox, radius, world, entity,
                entity.getPositionVector(),// new Vec3d(entity.motionX, entity.motionY, entity.motionZ),
                // isValid: should entity be pulled
                (Entity ent) -> (ent instanceof EntityItem) &&
                            (validMagnetII || Effects.doesMatch(sample,
                                    CompressedStack.createSampleStack(((EntityItem) ent).getItem())))
                ,
                // getSpeed: speed of the attraction
                null,
                // Post: When an entity is pulled
                null
        );

    }

    public static void onUpdateEntityItem(EntityCompressed entity, World world, ItemStack stack) {
        if (!CompressedStack.isCompressed(stack)) return;
        if (!Effects.canUseAny(stack, EnumEffect.ATTRACTION_I, EnumEffect.ATTRACTION_II, EnumEffect.ATTRACTION_III)) return;

        boolean canUseI = Effects.canUseAny(stack, EnumEffect.ATTRACTION_I);
        boolean canUseII = Effects.canUseAny(stack, EnumEffect.ATTRACTION_II);
        boolean canUseIII = Effects.canUseAny(stack, EnumEffect.ATTRACTION_III);

        ItemStack sample = CompressedStack.createSampleStack(stack);

        // radiusFactor is the tier of the stack - the lowest possible tier for a magnet
        float radiusFactor = CompressedStack.getTier(stack).ordinal() - EnumEffect.MAGNET_I.getTier().ordinal();
        float radius = 1.5f + radiusFactor;

        AxisAlignedBB boundingBox = entity.getEntityBoundingBox();
        AxisAlignedBB boundingBoxConsume = boundingBox.grow(0.7f);
        AxisAlignedBB boundingBoxEnts = boundingBoxConsume.grow(1f);
        Effects.attractEntities(boundingBox, radius, world, entity,
                entity.getPositionVector(),// new Vec3d(entity.motionX, entity.motionY, entity.motionZ),
                // isValid: should entity be pulled
                (Entity ent) -> {
                    if (ent.getEntityBoundingBox().intersects(boundingBoxConsume)) return false;
                    // Attraction I and II: attracting items
                    if (ent instanceof EntityItem) {
                        return canUseII || (canUseI && Effects.doesMatch(sample,
                                CompressedStack.createSampleStack(((EntityItem)ent).getItem())));
                    }
                    else {
                        if (ent.getEntityBoundingBox().intersects(boundingBoxEnts)) return false;
                        else if (ent instanceof EntityPlayer) {
                            return canUseIII && !ent.isSneaking() && !((EntityPlayer)ent).isCreative();
                        }
                        else {
                            return canUseIII;
                        }
                    }
                },
                // getSpeed: speed of the attraction
                null,
                // Post: When an entity is pulled
                null
        );

    }

}
